internal/gcs/guestconnection.go

```diff
@@ -234,31 +253,83 @@ func (gc *GuestConnection) newIoChannel() (*ioChannel, uint32, error) {
 	return newIoChannel(l), port, nil
 }
 
-func (gc *GuestConnection) requestNotify(cid string, ch chan struct{}) error {
+// requestNotify will use notifyCh to send notifications about container cid.
+// The GuestConnection will close notifyCh after notifying of container exit,
+// or upon close.
+//
+// Currently only OOM and Shutdown events are sent.
+func (gc *GuestConnection) requestNotify(cid string, notifyChs notifyChans) error {
 	gc.mu.Lock()
 	defer gc.mu.Unlock()
+
 	if gc.notifyChs == nil {
 		return errors.New("guest connection closed")
 	}
 	if _, ok := gc.notifyChs[cid]; ok {
 		return fmt.Errorf("container %s already exists", cid)
 	}
-	gc.notifyChs[cid] = ch
+
+	gc.notifyChs[cid] = notifyChs
+
 	return nil
 }
 
-func (gc *GuestConnection) notify(ntf *containerNotification) error {
-	cid := ntf.ContainerID
+func (gc *GuestConnection) notify(ntf *containerNotification) (err error) {
+	var (
+		cid    = ntf.ContainerID
+		ntType = ntf.Type
+		nntf   = nots.FromString(ntType)
+	)
+
+	entry := logrus.WithFields(logrus.Fields{
+		logfields.ContainerID: cid,
+		"notification-type":   nntf.String(),
+	})
+	entry.Debug("received notification from guest")
+
 	gc.mu.Lock()
-	ch := gc.notifyChs[cid]
-	delete(gc.notifyChs, cid)
+	ch, ok := gc.notifyChs[cid]
 	gc.mu.Unlock()
-	if ch == nil {
-		return fmt.Errorf("container %s not found", cid)
+	if !ok {
+		err = fmt.Errorf("container %s not found", cid)
+		entry.WithError(err).Error("could not notify appropriate container")
+		return
 	}
-	logrus.WithField(logfields.ContainerID, cid).Info("container terminated in guest")
-	close(ch)
-	return nil
+
+	// todo (helsaawy): internal/guest/prot/protocol has its own list of event types
+	// which doesnt match those in internal/notifications. internal/gcs/protocol
+	// has no corresponding list on the receiving end -- standardize all three
+	switch nntf {
+	case nots.ForcedExit, nots.GracefulExit, nots.UnexpectedExit:
+		// delete the entry
+		gc.mu.Lock()
+		delete(gc.notifyChs, cid)
+		gc.mu.Unlock()
+		// close the channels
+		ch.Close()
```

Comment:
It looks like there is a possible race condition here. If we receive OOM and exit events at roughly the same time, the exit event could close `ch`, and then the OOM event try to write to its channel, which will cause a panic.